llvm/docs/DirectX/DXILOpTableGenDesign.rst

```diff
@@ -93,76 +93,81 @@ attributes are specified as fields of the ``DXILOp`` class as described below.
+    class is described in a later section.
 
-A DXIL Operation is represented by the following TableGen class by encapsulating the various
-TableGen representations of its attributes described above.
+Types specific to DXIL
+----------------------
+
+Type notation used in this document viz., ``<size>Ty`` corresponds to TableGen records for
+LLVM types ``llvm_<size>_ty``. Apart from ``overloadTy`` described above, ``resRetF32Ty`` is
+used to denote resource return type and ``handleTy`` is used to denote handle type.
+
+Specification of DXIL Operation
+================================
+
+A DXIL Operation is represented by the following TableGen class that encapsulates the various
+TableGen representations of its properties described above.
 
 .. code-block::
 
    // Abstraction DXIL Operation
-   class DXILOp {
+   class DXILOp<int opcode, DXILOpClass opclass> {
      // A short description of the operation
      string Doc = "";
 
      // Opcode of DXIL Operation
-     int OpCode = 0;
+     int OpCode = opcode;
 
      // Class of DXIL Operation.
-     DXILOpClass OpClass = UnknownOpClass;
+     DXILOpClass OpClass = opclass;
 
      // LLVM Intrinsic DXIL Operation maps to
      Intrinsic LLVMIntrinsic = ?;
 
-     // Dag containing the arguments of the op. Default to 0 arguments.
-     dag arguments = (ins);
+     // Result type of the op.
+     list<LLVMType> result = [];
 
-     // Results of the op. Default to 0 results.
-     dag result = (out);
+     // List of argument types of the op. Default to 0 arguments.
+     list<LLVMType> arguments = [];
 
-     // List of constraints predicated on Shader Model version
-     list<SMVersionConstraints> sm_constraints;
-
-     // Non-predicated operation attributes
-     dag attrtibutes = (attrs);
-     Version DXILVersion = ?;
-   }
+     // List of valid overload types predicated by DXIL version
+     list<VersionedOverloads> overloads;
 
-Constraint Specification
-========================
+     // List of valid shader stages predicated by DXIL version
+    list<VersionedStages> stages;
 
-DXIL Operation attributes such as valid overload types and valid shader stages are
-predicated on Shader Model version. These are represented as list of constrained
-attributes.
-
-Following is the definition of a generic constraint and the associated predicate
-
-.. code-block::
+     // List of valid attributes predicated by DXIL version
+     list<VersionedAttributes> versioned_attributes = [];
+   }
 
-   // Primitive predicate
-   class Pred;
+Version Specification
+=====================
 
-   // Generic constraint
-   class Constraint<Pred pred> {
-     Pred predicate = pred;
-   }
+DXIL version is used to specify various version-dependent operation properties in
+place of Shader Model version.
 
-Shader Model version is represented as follows:
+A ``Version`` class encapsulating ``Major`` and ``Minor`` version number is defined
```

Comment:
> Should we call this `DXILVersion` instead of just `Version`? We have `DXILOp` above.

Since it is just a version, and has nothing DXIL-specific, I chose the class name with `DXIL`. The DXIL version records are defined as `DXIL1_*`.